src/integration.test.ts

```diff
@@ -58,58 +88,75 @@ describe("Integration Test", (): void => {
-        callback(null, { stdout, stderr });
-      }
-    );
-
-    await (load ? docker.loadDockerImages() : docker.saveDockerImages());
-
-    expect(core.getInput).nthCalledWith(1, "key", { required: true });
-    expect(core.info).nthCalledWith<[string]>(1, command);
-    expect(child_process.exec).lastCalledWith(
+  const assertExecBashCommand = (n: number, command: string): void => {
+    const infoN = 2 * n - 1; // core.info is called twice each time docker image list is called.
+    expect(core.info).nthCalledWith<[string]>(infoN, command);
+    expect(child_process.exec).nthCalledWith(
+      n,
       command,
       EXEC_OPTIONS,
       expect.anything()
     );
-    expect(core.info).lastCalledWith(stdout);
-    expect(core.error).lastCalledWith(stderr);
+    expect(core.error).lastCalledWith(STDERR);
     expect(core.setFailed).not.toHaveBeenCalled();
   };
 
+  const assertLoadDockerImages = (cacheHit: boolean): void => {
+    expect(cache.restoreCache).lastCalledWith([docker.DOCKER_IMAGES_PATH], KEY);
+    expect(core.saveState).nthCalledWith(1, docker.CACHE_HIT, cacheHit);
+    expect(core.setOutput).nthCalledWith(1, docker.CACHE_HIT, cacheHit);
+    if (!cacheHit) {
+      expect(core.saveState).lastCalledWith(
+        docker.DOCKER_IMAGES_LIST,
+        expect.anything()
+      );
+    }
+  };
+
+  const assertListAndLoad = (command: string): void => {
+    expect(core.getInput).nthCalledWith(1, "key", { required: true });
+    assertExecBashCommand(1, command);
+  };
+
   test("cache misses, then hits", async (): Promise<void> => {
     // Attempt first cache restore.
+    const listCommand =
+      'docker image list --format "{{ .Repository }}:{{ .Tag }}"';
     await docker.loadDockerImages();
+    assertLoadDockerImages(false);
+    assertListAndLoad(listCommand);
+    expect(core.info).lastCalledWith(dockerImages.join("\n"));
 
     // Expect cache miss since cache has never been saved.
-    expect(core.getInput).nthCalledWith(1, "key", { required: true });
     expect(core.setOutput).lastCalledWith(docker.CACHE_HIT, false);
-    expect(child_process.exec).not.toHaveBeenCalled();
-    expect(core.setFailed).not.toHaveBeenCalled();
     jest.clearAllMocks();
 
     // Run post step first time, expecting cache saved on cache miss.
-    const saveCommand =
-      'docker image list --format "{{ .Repository }}:{{ .Tag }}" | ' +
-      '2>&1 xargs --delimiter="\n" --no-run-if-empty --verbose --exit ' +
-      `docker save --output ${docker.DOCKER_IMAGES_PATH}`;
-    await mockedExec(false, saveCommand);
+    const saveCommand = `docker save --output ${docker.DOCKER_IMAGES_PATH} test-docker-image:v2`;
+    await docker.saveDockerImages();
+    expect(core.getState).nthCalledWith(1, docker.CACHE_HIT);
+    expect(core.getInput).lastCalledWith("read-only");
+    expect(core.getState).lastCalledWith(docker.DOCKER_IMAGES_LIST);
+    expect(cache.saveCache).toHaveBeenCalledTimes(1);
```

Comment:
In theory, this could be asserted about almost everything, but that seems like overkill. Is there a particular reason to use it here.